client/lib/cart/store/cart-synchronizer.js

```diff
@@ -52,11 +52,26 @@ function preprocessCartForServer( cart ) {
 		'products',
 		'coupon',
 		'is_coupon_applied',
+		'is_coupon_removed',
 		'currency',
 		'temporary',
 		'extra'
 	);
 
+	if (
+		! newCart.coupon &&
+		! newCart.is_coupon_applied &&
+		! newCart.is_coupon_removed &&
+		typeof document !== 'undefined'
+	) {
+		const coupon = url.parse( document.URL, true ).query.coupon;
+
+		if ( coupon ) {
+			newCart.coupon = coupon;
+			newCart.is_coupon_applied = false;
+		}
+	}
```

Comment:
we don't ned to block on this unless you want to update it, but we'd generally prefer the non-mutating variants of this pattern as well as avoiding `pick()`

```js
function preprocessCartForServer( fullCart ) {
	const {
		products,
		cloud,
		is_coupon_applied,
		is_coupon_removed,
		currency,
		temporary,
		extra,
	} = fullCart;

	// flipped booleans via _de Morgan's laws_ (pretty handy!)
	// https://en.wikipedia.org/wiki/De_Morgan%27s_laws
	const needsUrlCoupon = ! ( coupon || is_coupon_applied || is_coupon_removed );

	return {
		coupon,
		is_coupon_applied: needsUrlCoupon ? false : is_coupon_applied,
		is_coupon_removed,
		currency,
		temporary,
		extra,
		products: products.map(
			( { product_id, meta, free_trial, volume, extra } ) => ( { product_id, meta, free_trial, volume, extra } ) 
		)
	};
```

Reply:
I guess it's a matter of taste - I find the `condition && {}` version confusing and prefer more verbose and ugly imperative assignment

Reply:
it's a matter of taste to an extent. the imperative assignments create states that should never exist which opens the door for bugs. it also means that someone reading the code has to do more parsing and compiling in their head to make sure that they don't make improper assumptions about what the value of the binding is at a given point in time/in the lines.

you can clean it up in other ways too if your objection is ugly lines, but sometimes ugly lines are just what the doctor ordered because they aren't surprising - no "oh shoot, I _missed that_ that it was changing the value here"

```js
const cartProps = {
	coupon,
	is_coupon_applied,
	is_coupon_removed,
	currency,
	temporary,
	extra,
	products,
};

const urlCouponProps = needsUrlCoupon
	? { coupon: url.parse( document.URL, true ).query.coupon, is_coupon_applied: false }
	: {};

return {
	...cartProps,
	...urlCouponProps,
};
```

alternatively…

```js
…

return urlCouponProps
	? { ...cartProps, ...urlCouponProps }
	: cartProps;
```

alternatively…

```js
if ( ! needsUrlCoupon ) {
	return cartProps;
}

return {
	…cartProps,
	coupon: url.parse( document.URL, true ).query.coupon,
	is_coupon_applied: false,
};
```

that is, my objection isn't style - it's the risk of introducing hard-to-diagnose and easy-to-make bugs into the codebase that likely someone else will have to find 🤷‍♂️

Reply:
@dmsnell Good suggestion, thanks! I made this function more concise using some of your ideas: https://github.com/Automattic/wp-calypso/pull/25433/files#diff-5f9b5f654df0d1076ef22cec728a9451R25